src/Common/ThreadStatus.h

```diff
@@ -57,45 +57,57 @@ using ThreadStatusPtr = ThreadStatus *;
   * Create via CurrentThread::initializeQuery (for queries) or directly (for various background tasks).
   * Use via CurrentThread::getGroup.
   */
+class ThreadGroupStatus;
+using ThreadGroupStatusPtr = std::shared_ptr<ThreadGroupStatus>;
+
 class ThreadGroupStatus
 {
 public:
-    struct ProfileEventsCountersAndMemory
-    {
-        ProfileEvents::Counters::Snapshot counters;
-        Int64 memory_usage;
-        UInt64 thread_id;
-    };
-
-    mutable std::mutex mutex;
+    /// The first thread created this thread group
+    UInt64 master_thread_id = 0;
 
     ProfileEvents::Counters performance_counters{VariableContext::Process};
     MemoryTracker memory_tracker{VariableContext::Process};
 
-    ContextWeakPtr query_context;
-    ContextWeakPtr global_context;
+    /// Access to the members below has to be in critical section with  mutex
+    mutable std::mutex mutex;
```

Comment:
Usually, I would prefer quite the opposite logic: make read-only fields public and constant, and make mutable fields private and protected by a mutex. Ideally, it should be a class invariant so that it would be impossible to use mutable fields without the lock.